File: examples/test_o1_model_logic.js
#!/usr/bin/env node
/**
 * Simple test to verify the o1-mini model logic without dependencies.
 * This script tests the fix for issue #28 in JavaScript.
 */

function testO1ModelSystemMessageHandling() {
  console.log('🧪 Testing o1-mini model system message handling logic...');
  
  // Test data
  const testCases = [
    {
      model: 'o1-mini',
      message: 'What is 2+2?',
      systemMessage: 'You are a helpful math assistant.',
      expectedSystem: null,
      description: 'o1-mini should convert system message to user message prefix'
    },
    {
      model: 'o1-preview', 
      message: 'Hello world',
      systemMessage: 'You are helpful.',
      expectedSystem: null,
      description: 'o1-preview should convert system message to user message prefix'
    },
    {
      model: 'o3-mini',
      message: 'Test message', 
      systemMessage: 'System prompt',
      expectedSystem: null,
      description: 'o3-mini should convert system message to user message prefix'
    },
    {
      model: 'gpt-4o',
      message: 'What is 2+2?',
      systemMessage: 'You are a helpful math assistant.',
      expectedSystem: 'You are a helpful math assistant.',
      description: 'Regular models should preserve system messages'
    },
    {
      model: 'claude-3-opus',
      message: 'Hello',
      systemMessage: 'Be helpful',
      expectedSystem: 'Be helpful', 
      description: 'Non-o1 models should preserve system messages'
    }
  ];
  
  console.log('\n📝 Running test cases...');
  
  testCases.forEach((testCase, i) => {
    console.log(`\n${i + 1}. ${testCase.description}`);
    console.log(`   Model: ${testCase.model}`);
    console.log(`   Original message: ${testCase.message}`);
    console.log(`   Original system: ${testCase.systemMessage}`);
    
    // Apply the fix logic
    let message = testCase.message;
    let systemMessage = testCase.systemMessage;
    const gptModel = testCase.model;
    
    // O1-series models don't support system messages
    const o1Models = ['o1-mini', 'o1-preview', 'o3-mini'];
    
    if (o1Models.includes(gptModel)) {
      // For o1-series models, prepend system message to user message instead
      if (systemMessage && systemMessage.trim()) {
        message = `System instructions: ${systemMessage}\n\nUser message: ${message}`;
      }
      systemMessage = null; // Don't send system message for o1 models
    }
    
    console.log(`   Final message: ${message}`);
    console.log(`   Final system: ${systemMessage}`);
    
    // Verify the result
    if (systemMessage === testCase.expectedSystem) {
      if (o1Models.includes(gptModel)) {
        const expectedPrefix = `System instructions: ${testCase.systemMessage}`;
        if (message.includes(expectedPrefix)) {
          console.log('   ✅ PASS - System message correctly converted to user message prefix');
        } else {
          console.log('   ❌ FAIL - System message not properly converted');
        }
      } else {
        console.log('   ✅ PASS - System message correctly preserved');
      }
    } else {
      console.log(`   ❌ FAIL - Expected system: ${testCase.expectedSystem}, got: ${systemMessage}`);
    }
  });
  
  console.log('\n🎉 Test completed!');
  console.log('🔧 The fix properly handles o1-series models by:');
  console.log('   • Converting system messages to user message prefixes for o1-mini, o1-preview, o3-mini');
  console.log('   • Preserving system messages for all other models');
  console.log('   • Avoiding the \'messages[0].role\' does not support \'system\' error');
}

if (require.main === module) {
  testO1ModelSystemMessageHandling();
}

File: examples/test_o1_model_logic.py
#!/usr/bin/env python3
"""
Simple test to verify the o1-mini model logic without dependencies.
This script tests the fix for issue #28.
"""


def test_o1_model_system_message_handling():
    """Test that o1-mini model properly handles system messages."""
    
    print("🧪 Testing o1-mini model system message handling logic...")
    
    # Test data
    test_cases = [
        {
            "model": "o1-mini",
            "message": "What is 2+2?",
            "system_message": "You are a helpful math assistant.",
            "expected_system": None,
            "description": "o1-mini should convert system message to user message prefix"
        },
        {
            "model": "o1-preview", 
            "message": "Hello world",
            "system_message": "You are helpful.",
            "expected_system": None,
            "description": "o1-preview should convert system message to user message prefix"
        },
        {
            "model": "o3-mini",
            "message": "Test message", 
            "system_message": "System prompt",
            "expected_system": None,
            "description": "o3-mini should convert system message to user message prefix"
        },
        {
            "model": "gpt-4o",
            "message": "What is 2+2?",
            "system_message": "You are a helpful math assistant.",
            "expected_system": "You are a helpful math assistant.",
            "description": "Regular models should preserve system messages"
        },
        {
            "model": "claude-3-opus",
            "message": "Hello",
            "system_message": "Be helpful",
            "expected_system": "Be helpful", 
            "description": "Non-o1 models should preserve system messages"
        }
    ]
    
    print("\n📝 Running test cases...")
    
    for i, case in enumerate(test_cases, 1):
        print(f"\n{i}. {case['description']}")
        print(f"   Model: {case['model']}")
        print(f"   Original message: {case['message']}")
        print(f"   Original system: {case['system_message']}")
        
        # Apply the fix logic
        message = case['message']
        system_message = case['system_message']
        gpt_model = case['model']
        
        # O1-series models don't support system messages
        o1_models = ['o1-mini', 'o1-preview', 'o3-mini']
        
        if gpt_model in o1_models:
            # For o1-series models, prepend system message to user message instead
            if system_message and system_message.strip():
                message = f"System instructions: {system_message}\n\nUser message: {message}"
            system_message = None  # Don't send system message for o1 models
        
        print(f"   Final message: {message}")
        print(f"   Final system: {system_message}")
        
        # Verify the result
        if system_message == case['expected_system']:
            if gpt_model in o1_models:
                expected_prefix = f"System instructions: {case['system_message']}"
                if expected_prefix in message:
                    print(f"   ✅ PASS - System message correctly converted to user message prefix")
                else:
                    print(f"   ❌ FAIL - System message not properly converted")
            else:
                print(f"   ✅ PASS - System message correctly preserved")
        else:
            print(f"   ❌ FAIL - Expected system: {case['expected_system']}, got: {system_message}")
    
    print("\n🎉 Test completed!")
    print("🔧 The fix properly handles o1-series models by:")
    print("   • Converting system messages to user message prefixes for o1-mini, o1-preview, o3-mini")
    print("   • Preserving system messages for all other models")
    print("   • Avoiding the 'messages[0].role' does not support 'system' error")


if __name__ == "__main__":
    test_o1_model_system_message_handling()